### ISSUE TITLE
Prototype support for list queries in Intermine data-source

### ISSUE THREAD

**Issue by alancleary** [legumeinfo/graphql-server#97]
The use-case is to support pan-gene list queries.

**Comment by sammyjava** [comment]
Turns out it indeed can be done, here's an XML example (run against MiniMine):
```xml
<query name="" model="genomic" view="Gene.primaryIdentifier" longDescription="">
  <constraint path="Gene" op="IN" ids="39481717, 39498180, 39500146, 39505454, 39520896"/>
</query>
```
Note that you have to supply a list of IDs, not the persistent identifiers like `primaryIdentifier`. You can use the `IdsConstraint` object in imjs as well: http://intermine.org/imjs/#http://intermine.org/imjs/class/IdsContraint.html.

There isn't a way to create a query like this using the Query Builder since you're not supposed to work with IDs. But it's in the API and you can enter the above XML and it works.

**Comment by sammyjava** [comment]
A prerequisite on this is to translate a provided set of identifiers to IDs on the mine before executing the query on the list of IDs. So it's a two-step process, which I have no idea how to implement using Promises. The query is easy:
```xml
<query name="" model="genomic" view="Gene.id" longDescription="">
  <constraint path="Gene.primaryIdentifier" op="=" value="cicar.ICC4958.gnm2.ann1.Ca_00666"/>
</query>
```
but you have to execute them one at a time.

**Comment by adf-ncgr** [comment]
I think we'd probably want to go this route:
http://intermine.org/imjs/#http://intermine.org/imjs/class/Service.html#resolveIds-dynamic
(he says without entirely understanding how to do it)

**Comment by alancleary** [comment]
> I think we'd probably want to go this route: http://intermine.org/imjs/#http://intermine.org/imjs/class/Service.html#resolveIds-dynamic (he says without entirely understanding how to do it)

Hopefully not. This submits a job to Intermine so you need to poll the server to see if the job has completed and then fetch the result manually. Like @sammyjava said at today's stand-up, ideally the solution will be stateless!

**Comment by adf-ncgr** [comment]
well, I'm dubious that this will scale well but the form you were suggesting would I think be something like this:
```
<query name="" model="genomic" view="Gene.id" longDescription="" constraintLogic="A or B or C">
  <constraint path="Gene.primaryIdentifier" code="A" op="=" value="cicar.ICC4958.gnm2.ann1.Ca_00666"/>
  <constraint path="Gene.primaryIdentifier" code="B" op="=" value="cicar.ICC4958.gnm2.ann1.Ca_00667"/>
  <constraint path="Gene.primaryIdentifier" code="C" op="=" value="cicar.ICC4958.gnm2.ann1.Ca_00668"/>
</query>
```

**Comment by alancleary** [comment]
There's that constraint logic again. Good thing we added it! If performance is a concern we could put a limit on the number of IDs, like 100 or whatever seems reasonable.

**Comment by adf-ncgr** [comment]
I'm not %100 sure but in playing a bit with the interface it looks like Intermine might think that more than 26 constraints (A-Z) is unreasonable; which seems unreasonably limiting to me. This conclusion is based on various errors I see when trying to add one more constraint above the apparently alphabetical limit. I'd rather be stateful than constrained that much! I suppose that if the intermine API doesn't conveniently support what we want for the first part (conversion from public ids to internal ids) we could always write our own behind-the-scenes service that talks to the same underlying db (or an alternate representation of data sucked from that underlying db). But I'd rather we work with the existing implementation if possible.

**Comment by adf-ncgr** [comment]
PS. this error message seems to confirm my interpretation of the limit:

> Import of query(ies) failed. Invalid format of query: "The constraint code must be a single plain latin uppercase character"

**Comment by alancleary** [comment]
Thanks for playing around with that @adf-ncgr. Regarding alternatives, I did have the thought that this is the sort of thing Redis handles very well (and fast!). Just a thought.

**Comment by adf-ncgr** [comment]
Ironically, I had thought you kids were going to be bringing up redis as a solution to the original pangene translation use case...

**Comment by alancleary** [comment]
Redis would certainly be a good solution from a technology perspective, but we would have to develop it from scratch so I think Intermine and BDB should definitely be pursued first.

**Comment by sammyjava** [comment]
In the Java API you can supposedly use
http://intermine.org/intermine/org/intermine/pathquery/Constraints.html#oneOfValues(java.lang.String,java.util.Collection)
to provide a list of values (primary identifiers, say) in a query *(that returns IDs, say).

In imjs the equivalent seems to be
http://intermine.org/imjs/#http://intermine.org/imjs/class/MultiValueConstraint.html
to provide the array of String values.

HOWEVER, I cannot seem to use PathQuery.toXml() to see what the Java query would look like in XML - it returns blank. And if I try to send the query to the InterMine service using the Java API it complains that the query is blank - because it uses toXml() to convert it as well. So it may not really be supported, but maybe the imjs version works.

**Comment by sammyjava** [comment]
Never mind, I'm doing something wrong, a plain old EQ query is also bombing in my test program. I'll get the XML version....

**Comment by adf-ncgr** [comment]
well that sounds promising! If it works as advertising it sounds like it would actually obviate the need to deal with the internal ids at all since you could just use the multivalue constraint on primaryIdentifiers as part of the query of primary interest, where I had initially used a List. I can't seem to figure out how to make the QueryBuilder construct such a thing, but sounds like you are on the right track

**Comment by sammyjava** [comment]
Correctamundo, back to a single query which I can write. The webapp Query Builder doesn't support these, but there are a lot of path queries that aren't supported by the Query Builder. My issue is that they've totally rearranged the standalone Java webservice client API (I haven't used it in ages and hardly ever use it) so I'm wrangling my little package to use the new package layout. Presumably something changed in 5.0.+ on the server side that makes it think I'm sending empty queries with the old client API.

**Comment by sammyjava** [comment]
Yup, I was hitting a new mine with an old API. Here's the skinny, easy-peasy:

PathQuery( view: [Gene.id, Gene.name, Gene.length], orderBy: [], constraints: {Gene.primaryIdentifier ONE OF [phavu.G19833.gnm1.ann1.Phvul.011G091200, phavu.G19833.gnm1.ann1.Phvul.005G002800, phavu.G19833.gnm1.ann1.Phvul.005G149000, phavu.G19833.gnm1.ann1.Phvul.008G102000]=A}, logic: A, outerJoinStatus: {}, descriptions: {}, description: null)
```xml
<query name="query" model="genomic" view="Gene.id Gene.name Gene.length" longDescription=""><constraint path="Gene.primaryIdentifier" op="ONE OF"><value>phavu.G19833.gnm1.ann1.Phvul.011G091200</value><value>phavu.G19833.gnm1.ann1.Phvul.005G002800</value><value>phavu.G19833.gnm1.ann1.Phvul.005G149000</value><value>phavu.G19833.gnm1.ann1.Phvul.008G102000</value></constraint></query>
```
Returns:
```
41513542	Phvul.005G149000	2277
41588546	Phvul.011G091200	1194
41588556	Phvul.005G002800	2795
41588566	Phvul.008G102000	10085
```

**Comment by sammyjava** [comment]
That XML actually works if you punch it into the Query Builder, you just can't build it from scratch AFAICT.

**Comment by adf-ncgr** [comment]
Great, I thought I'd tried something basically similar to that but I must have had it slightly wrong. Taking your basic model for this and adapting it to the earlier pangenes query where I was using a list, I find that this works:
```
<queries><query name="" model="genomic" view="Gene.primaryIdentifier Gene.name Gene.assemblyVersion Gene.annotationVersion Gene.description Gene.organism.name Gene.strain.identifier Gene.panGeneSets.genes.primaryIdentifier Gene.panGeneSets.genes.name Gene.panGeneSets.genes.assemblyVersion Gene.panGeneSets.genes.annotationVersion Gene.panGeneSets.genes.description Gene.panGeneSets.genes.organism.name Gene.panGeneSets.genes.strain.identifier" longDescription="" sortOrder="Gene.primaryIdentifier asc" constraintLogic="B and A">
  <constraint path="Gene.primaryIdentifier" code="B" op="CONTAINS" value="glyma.Wm82.gnm4.ann1"/>
  <constraint path="Gene.panGeneSets.genes.primaryIdentifier" code="A" op="ONE OF"> <value>glyma.Lee.gnm2.ann1.Gm_00467</value>
<value>glyma.Lee.gnm2.ann1.Gm_06449</value>
<value>glyma.Lee.gnm2.ann1.Gm_16589</value>
</constraint>
</query></queries>
```
I'll try to do some additional testing of the approach to see how it scales to larger input lists.

**Comment by sammyjava** [comment]
FYI, indexing should make the Gene.primaryIdentifier part go faster if you use "=" instead of "CONTAINS" on Gene.organism.abbreviation="glyma", gene.strain.identifier="Wm82", gene.assemblyVersion="gnm4", and gene.annotationVersion="ann1". Of course that's four, not one, constraintLogic elements, but that 's fine.
```xml
<query name="" model="genomic" view="Gene.primaryIdentifier Gene.name Gene.description Gene.strain.identifier Gene.panGeneSets.genes.primaryIdentifier" longDescription="" sortOrder="Gene.primaryIdentifier asc" constraintLogic="A and B and C and D and E">
  <constraint path="Gene.panGeneSets.genes.primaryIdentifier" code="A" op="ONE OF">
    <value>glyma.Lee.gnm2.ann1.Gm_00467</value>
    <value>glyma.Lee.gnm2.ann1.Gm_06449</value>
    <value>glyma.Lee.gnm2.ann1.Gm_16589</value>
  </constraint>
  <constraint path="Gene.organism.abbreviation" code="B" op="=" value="glyma"/>
  <constraint path="Gene.annotationVersion" code="C" op="=" value="ann1"/>
  <constraint path="Gene.assemblyVersion" code="D" op="=" value="gnm4"/>
  <constraint path="Gene.strain.identifier" code="E" op="=" value="Wm82"/>
</query>
```

**Comment by adf-ncgr** [comment]
Yeah, I was being a bit lazy; I actually meant to use "LIKE 'prefix%'" so it could take advantage of indexing, but I guess LIKE is only supported in data tables and not in path queries? In any case, your version is probably better.

**Comment by alancleary** [comment]
> FYI, indexing should make the Gene.primaryIdentifier part go faster if you use "=" instead of "CONTAINS" on Gene.organism.abbreviation="glyma", gene.strain.identifier="Wm82", gene.assemblyVersion="gnm4", and gene.annotationVersion="ann1". Of course that's four, not one, constraintLogic elements, but that 's fine.
> 
> ```
> <query name="" model="genomic" view="Gene.primaryIdentifier Gene.name Gene.description Gene.strain.identifier Gene.panGeneSets.genes.primaryIdentifier" longDescription="" sortOrder="Gene.primaryIdentifier asc" constraintLogic="A and B and C and D and E">
>   <constraint path="Gene.panGeneSets.genes.primaryIdentifier" code="A" op="ONE OF">
>     <value>glyma.Lee.gnm2.ann1.Gm_00467</value>
>     <value>glyma.Lee.gnm2.ann1.Gm_06449</value>
>     <value>glyma.Lee.gnm2.ann1.Gm_16589</value>
>   </constraint>
>   <constraint path="Gene.organism.abbreviation" code="B" op="=" value="glyma"/>
>   <constraint path="Gene.annotationVersion" code="C" op="=" value="ann1"/>
>   <constraint path="Gene.assemblyVersion" code="D" op="=" value="gnm4"/>
>   <constraint path="Gene.strain.identifier" code="E" op="=" value="Wm82"/>
> </query>
> ```
> 
> It seems to run much faster, but that may be database caching.

Is this what we want to implement? I could knock this out pretty easily.

**Comment by adf-ncgr** [comment]
Yeah, I think we should go ahead with one of these versions. FWIW, I did some testing on a large input list (~2700 genes) and it actually seems like the version I had originally formulated is faster (2 sec vs 8 sec, consistently after repeated attempts to try to counteract effects of caching). Maybe extra joins required in @sammyjava version are slower? Not sure my test is entirely conclusive, but I guess it shouldn't be too hard to switch between these two fairly similar formulations later on if needed.

**Comment by sammyjava** [comment]
Well, regardless of speed, I say that if the user is entering a strain (Wm82), an assembly version (gnm4), an annotation version (ann2) on the UI, we should do an "=" constraint on those terms rather than collect them into a string (glyma.Wm82.gnm4.ann1) that we then run a CONTAINS on, even if some testing shows it to be faster. (My test was quite faster with the "=" JOINS with a short list of genes, and a test of 2700 genes is NOT a typical test case for optimization, albeit a typical test case for ADF usage. Apple is still trying to figure out how to make MacOS support 2700 open apps.)

**Comment by sammyjava** [comment]
> This is interesting because I was under the impression that `AND` constraint logic is used by default when no logic is specified. Did you confirm that the two queries produce the same result?

AND is definitely the default when constraintLogic is not supplied. It's what we see all the time in the Query Builder, although sometimes it shows a comma rather than "AND". But the XML is AND.

**Comment by sammyjava** [comment]
BTW I'm happy to hit this now that it's a straightforward task, still slower than if @alancleary did it but I'm free-ish on LIS stuff right now.

**Comment by adf-ncgr** [comment]
fine by me- is there a natural graphql analog to what we will be translating into this intermine pathquery construct or is that something that needs further discussion?

**Comment by sammyjava** [comment]
To annoyingly belabor but make a design point: graphql-server doesn't and shouldn't "know" that a primary identifier contains gensp.Strain.gnm.ann in order to form a query. There is nothing in graphql-server that tells it what a primary identifier looks like and we don't parse them. However, Types that extend SequenceFeature all have the needed bits for the "=" constraints on those bits.

**Comment by sammyjava** [comment]
> fine by me- is there a natural graphql analog to what we will be translating into this intermine pathquery construct or is that something that needs further discussion?

I think it's just like any of the other queries written in the data-sources/intermine/api methods. You get parameters and you run a query and you return the results. Unless I'm missing something.

**Comment by alancleary** [comment]
> Well, regardless of speed, I say that if the user is entering a strain (Wm82), an assembly version (gnm4), an annotation version (ann2) on the UI, we should do an "=" constraint on those terms rather than collect them into a string (glyma.Wm82.gnm4.ann1) that we then run a CONTAINS on, even if some testing shows it to be faster.
> ...
> To annoyingly belabor but make a design point: graphql-server doesn't and shouldn't "know" that a primary identifier contains gensp.Strain.gnm.ann in order to form a query. There is nothing in graphql-server that tells it what a primary identifier looks like and we don't parse them. However, Types that extend SequenceFeature all have the needed bits for the "=" constraints on those bits.

I didn't catch that these parameters were being cobbled together into a specific ID structure. In this case I agree that we definitely shouldn't do that. Let's keep our coupling nice and loose :wink:

> > fine by me- is there a natural graphql analog to what we will be translating into this intermine pathquery construct or is that something that needs further discussion?
> 
> I think it's just like any of the other queries written in the data-sources/intermine/api methods. You get parameters and you run a query and you return the results. Unless I'm missing something.

Pretty much. The only difference here is that one of the parameters is an array, which I believe is a first for our GraphQL API. Feel free to tackle it @sammyjava. It sounds like you can get to it sooner than me.

**Comment by adf-ncgr** [comment]
Since we're belaboring the point, I would add that we need to ensure that a UI with separate selectors for the component bits will need to ensure that the user can't request nonsensical combinations. Maybe this can be done in an analogous way to how selecting a genus constrains the set of species choices (but we need to also constrain gnm and ann versions here). Nevertheless, I would agree that doing things this way is more combinatorially elegant than having a long dropdown of full yuck prefixes.

Agreed that having a List (or array) as a parameter makes sense here. I still get a little hung up on the fact that GraphQL is more like an API and less like a query language in regards to how its inputs are handled (e.g. the choice of whether to translate a String parameter into an equals or a contains constraint is a decision made internally in the implementation).

**Comment by alancleary** [comment]
> Since we're belaboring the point, I would add that we need to ensure that a UI with separate selectors for the component bits will need to ensure that the user can't request nonsensical combinations. Maybe this can be done in an analogous way to how selecting a genus constrains the set of species choices (but we need to also constrain gnm and ann versions here). Nevertheless, I would agree that doing things this way is more combinatorially elegant than having a long dropdown of full yuck prefixes.

This sounds like the type of thing that should be in the UI spec!

**Comment by adf-ncgr** [comment]
good call, I'll refresh my memory as to what @StevenCannon-USDA has started for this and propose changes to that as seems appropriate.

**Comment by sammyjava** [comment]
> Since we're belaboring the point, I would add that we need to ensure that a UI with separate selectors for the component bits will need to ensure that the user can't request nonsensical combinations.

Good point. Fortunately for me, I'm writing the GraphQL implementation, not the web component. :) (I've pretty much stayed away from web component coding for quite a while.)

**Comment by sammyjava** [comment]
Thinking out loud, for UI purposes I guess we'd implement a getStrainAssemblyVersions API method (give it a Strain, get back a list of assemblyVersions) and a getAssemblyAnnotationVersions API method (give it a Strain and assemblyVersion and get back a list of annotationVersions). Pretty easy to do to make those selectors populate correctly in order as will most certainly be required by the UI spec! :)

**Comment by adf-ncgr** [comment]
sounds good although I'm not sure we ought to assume that Strains are unique across species (not that I have any actual examples at hand but it wouldn't surprise me if glyma.Davis met a phalu.Davis someday). ie genus and species probably ought to be included in these signatures as well.

**Comment by sammyjava** [comment]
This is downstream from selecting genus and species.

**Comment by alancleary** [comment]
> Thinking out loud, for UI purposes I guess we'd implement a getStrainAssemblyVersions API method (give it a Strain, get back a list of assemblyVersions) and a getAssemblyAnnotationVersions API method (give it a Strain and assemblyVersion and get back a list of annotationVersions). Pretty easy to do to make those selectors populate correctly in order as will most certainly be required by the UI spec! :)

By "API method" do you mean in the GraphQL API? I would be inclined to keep such a specific endpoint out of the API in preference for getting those data through an appropriately structured query.

**Comment by sammyjava** [comment]
I mean just like how we get anything else from the methods under `src/data-sources/intermine/api/` which I call the "API methods." If you're thinking it should be done differently, then I throw the ball in your half-court, because that's the only way I know how to do it.
`

**Comment by sammyjava** [comment]
Something like `get-pangene-homologs.ts` or something. Give it a list of genes and the other stuff, get back paginated results containing genes.

**Comment by alancleary** [comment]
Ah. Let's take a step back here; I got swept away by the conversation. The point of this issue is to add support for list queries to the Intermine data source, not to implement a specific list query. That should wait until a UI spec is in place. For now all that needs to be implemented is an `intermineOneOfConstraint` function in the `src/data-sources/intermine/intermine.server.ts` file.

**Comment by sammyjava** [comment]
Ahhh, right. We're at a higher level. I got thrown off by "The use-case is to support pan-gene list queries" which we won't be doing here, just making it possible.

I forget that we code each constraint op like "ONE OF" as a separate thing in `src/data-sources/intermine/intermine.server.ts`.

Which you'll do, not me - I'll stay off of `src/data-sources/intermine/intermine.server.ts` and remain a user of that code.

**Comment by sammyjava** [comment]
OK, I've got some time on my hands so I'll give implementing `intermineOneOfConstraint` the ol' college try, followed by the example of genes in the same pan-gene sets as the supplied list of genes (for a given gensp, Strain, ann, gnm).

**Comment by sammyjava** [comment]
OK I knew this was going to be hard for me. I have absolutely no idea how to iterate through the supplied values (gene identifiers in this case) within the constraint, the way it's written. Here's I mean with an array of three elements:
```typescript
export const intermineOneOfConstraint =
    (path: string, values: Array<string>, code: string=''): string => {
        const codeAttr = code ? `code='${code}'` : '';
        return `<constraint path='${path}' ${codeAttr} op='ONE OF'>
   <value>values[0]</value>
   <value>values[1]</value>
   <value>values[2]</value>
</constraint>`;
    };
```
If you could supply that piece of code I think I can keep moving.

**Comment by alancleary** [comment]
> OK I knew this was going to be hard for me. I have absolutely no idea how to iterate through the supplied values (gene identifiers in this case) within the constraint, the way it's written. Here's I mean with an array of three elements:
> 
> ```ts
> export const intermineOneOfConstraint =
>     (path: string, values: Array<string>, code: string=''): string => {
>         const codeAttr = code ? `code='${code}'` : '';
>         return `<constraint path='${path}' ${codeAttr} op='ONE OF'>
>    <value>values[0]</value>
>    <value>values[1]</value>
>    <value>values[2]</value>
> </constraint>`;
>     };
> ```
> 
> If you could supply that piece of code I think I can keep moving.

Looks like an opportunity for a [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) and [join](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join) combo. These methods are already being used in the file you're working in: `src/data-sources/intermine/intermine.server.ts`

**Comment by sammyjava** [comment]
Those don't prepend the `<value>` and postpend (?) `</value>` on each element. How about you just write it for me and I'll keep going. I never signed up to learn Typescript at this job, TBH.

**Comment by adf-ncgr** [comment]
I think I more or less grok what Alan is suggesting, but being horrible at javascript myself (not to mention typescript), I will try to send something that may give you a lightbulb in a few minutes. It will undoubtedly smell a bit like perl...

**Comment by adf-ncgr** [comment]
something like this? 
`Array.from(Array(values.length).keys()).map((x) => '<value>values['+x+']</value>').join("\n")`
took me longest to figure out how to do a sequence of numbers in javascript- pretty hideous! maybe there's a better way for that, but I think the map/join part is more or less what @alancleary was suggesting?

**Comment by alancleary** [comment]
> `Array.from(Array(values.length).keys()).map((x) => '<value>values['+x+']</value>').join("\n")`

:scream:

**Comment by adf-ncgr** [comment]
Actually, I may have misunderstood slightly what was needed here, we need to interpolate the actual values into the return string, right? If so, then move the quotes a bit. Thanks for the vote of horrified confidence there, Alan. I told you it would smell perly.

**Comment by alancleary** [comment]
```js
const valueTags = values.map((v) => `<value>${v}</value>`).join('');
```

**Comment by alancleary** [comment]
> Actually, I may have misunderstood slightly what was needed here, we need to interpolate the actual values into the return string, right? If so, then move the quotes a bit. Thanks for the vote of horrified confidence there, Alan. I told you it would smell perly.

The important thing is that you tried :wink:

**Comment by adf-ncgr** [comment]
D'oh! I was so intent on getting the string values[0] in there I missed that much better approach.

**Comment by adf-ncgr** [comment]
> The important thing is that you tried 😉

Or is the important thing that I managed to fish out of you what @sammyjava wanted? In any case, I hope you won't start calling me "Patches" now...

**Comment by sammyjava** [comment]
> The important thing is that you tried 😉

@adf-ncgr will try anything.... :)

**Comment by sammyjava** [comment]
OK, I've implemented it and it tests successfully with my proposed getPanGenesOfGenes method. It's in this new branch: https://github.com/legumeinfo/graphql-server/tree/one-of-constraint
I'll wait on a PR for @alancleary 's review of the `intermineOnOfConstraint` implementation which is the main thing. But I needed to implement something that uses it to test it. I can yank the getPanGenesOfGenes example from the PR if we want to keep them separate.

I noticed one important thing testing this example against MiniMine: since we query all attributes of Gene (`intermineGeneAttributes`), MiniMine only returns those genes that are fully populated (which are only a couple of annotations), not the other pan-genes that just have identifiers loaded from the pan-gene set collection. I don't think this is an issue in production mines, but it's a bit of a shocker when testing against MiniMine. I generally load all relevant annotations in the production mines so it's just a MiniMine issue.
